### PR TITLE
Rewrite statsd metrics reporter to support kafka 0.8.2

### PR DESCRIPTION
**Pull request by jun-he**
cc/ @liukai @alexism 

I wrote a datadog statsd metrics reporter for kafka 0.8.2.  The new reporter will send metrics with tags to  statsd.  The tags are extracted from metrics' tags generated by kafka 0.8.2.  It will make it is easier for us to check kafka metrics in statsd.

I appreciate if you can help to review it. 

Thanks!

Jun

### CODE REVIEW

**Comment by alexism** [comment]
All files must have the apache license header.

File: build.gradle [hunk 1]
```diff
@@ -32,29 +14,30 @@ apply plugin: 'idea'
 apply plugin: 'com.github.johnrengelman.shadow'
 
 group = 'com.airbnb'
-description = 'StatsD Support for Kafka Metrics'
```
Comment (alexism) [inline build.gradle:35]: This is not specific to datadog, is it?

File: build.gradle [hunk 2]
```diff
@@ -32,29 +14,30 @@ apply plugin: 'idea'
...
 
 repositories {
     mavenCentral()
-    maven { url 'http://dl.bintray.com/readytalk/maven' }
+    maven { url "http://dl.bintray.com/airbnb/jars" }
 }
+
 dependencies {
-    compile('org.apache.kafka:kafka_2.10:0.8.1.1')
-    compile 'log4j:log4j:+'
```
Comment (alexism) [inline build.gradle:43]: don't we need a dependency on `com.indeedeng:java-statsd-client:2.0.7` ?

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java [hunk 3]
```diff
@@ -23,119 +21,130 @@
...
- *
- * The reporter is enabled and configured from the Kafka config file, server.properties.
  *
- * @see kafka.metrics.KafkaMetricsReporter
  */
 public class KafkaStatsdMetricsReporter implements KafkaStatsdMetricsReporterMBean, KafkaMetricsReporter {
 
+    private static final org.slf4j.Logger log = LoggerFactory.getLogger(StatsDReporter.class);
 
-    public static final String DEFAULT_EXCLUDE_REGEX = "(kafka\\.consumer\\.FetchRequestAndResponseMetrics.*)|(.*ReplicaFetcherThread.*)|(kafka\\.server\\.FetcherLagMetrics\\..*)|(kafka\\.log\\.Log\\..*)|(kafka\\.cluster\\.Partition\\..*)";
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java:48]: would you exclude or include the matches? having `EXCLUDE`  in the name is unambiguous.

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java [hunk 4]
```diff
@@ -23,119 +21,130 @@
...
+    public static final String DEFAULT_REGEX_FILTER = "(kafka\\.consumer\\.FetchRequestAndResponseMetrics.*)|(.*ReplicaFetcherThread.*)|(kafka\\.server\\.FetcherLagMetrics\\..*)|(kafka\\.log\\.Log\\..*)|(kafka\\.cluster\\.Partition\\..*)";
 
-    private AbstractPollingReporter underlying;
-    private boolean enabled = false;
-    private boolean running = false;
-    private final Logger logger = Logger.getLogger(getClass());
+    @Override
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: methods should be after attributes.

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java [hunk 5]
```diff
@@ -23,119 +21,130 @@
...
-        metricPredicate = MetricPredicate.ALL;
-        metricDimensionOptions = MetricDimensionOptions.fromProperties(props.props(), "external.kafka.statsd.dimension.enabled.");
-
-        String exclude_regex = props.getString("external.kafka.statsd.metrics.exclude_regex", DEFAULT_EXCLUDE_REGEX);
-        if (exclude_regex.length() != 0){
-            metricPredicate = new ExcludeMetricPredicate(exclude_regex);
+        prefix = props.getString("external.kafka.statsd.metrics.prefix", "");
+        pollingPeriodInSeconds = props.getInt("kafka.metrics.polling.interval.secs", 10);
+        metricDimensionOptions = MetricDimensionOptions.fromProperties(props.props()
+                , "external.kafka.statsd.dimension.enabled.");
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: comma should be at the end of the line

**Comment by alexism** [comment]
Let's follow https://google-styleguide.googlecode.com/svn/trunk/javaguide.html

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java [hunk 6]
```diff
@@ -23,119 +21,130 @@
...
-            if (logger.isInfoEnabled())
-                logger.info("Kafka Statsd metrics reporter is enabled");
-
-            startReporter(pollingIntervalSecs);
-        } else {
-            if (logger.isInfoEnabled())
-                logger.info("Kafka Statsd metrics reporter is disabled");
-
-        }
+        this.isTagSupported = props.getBoolean("external.kafka.statsd.support.tag", true);
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: I think `enabled` is more accurate than `supported`.
Can we rename the variable into `isTagEnabled` and the property to `external.kafka.statsd.tag.enabled`

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java [hunk 7]
```diff
@@ -23,119 +21,130 @@
...
+        }
+    }
+
+    private void loadConfig(VerifiableProperties props) {
         enabled = props.getBoolean("external.kafka.statsd.reporter.enabled", false);
-        int pollingIntervalSecs = props.getInt("kafka.metrics.polling.interval.secs", 10);
         host = props.getString("external.kafka.statsd.host", "localhost");
         port = props.getInt("external.kafka.statsd.port", 8125);
-        prefix = props.getString("external.kafka.statsd.metrics.prefix", null);
-        metricPredicate = MetricPredicate.ALL;
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java:77]: `metricPredicate` does not seem to have a default value anymore.
Let's stay explicit and default to `MetricPredicate.ALL`

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java [hunk 8]
```diff
@@ -23,119 +21,130 @@
...
-            if (logger.isInfoEnabled())
-                logger.info(String.format("Started Kafka Statsd metrics reporter with host=%s, port=%d, polling_period_secs=%d, prefix=%s", host, port, pollingPeriodSecs, prefix == null ? "" : prefix));
-        } else {
-            if (logger.isInfoEnabled())
-                logger.info("Kafka Statsd metrics reporter is already running");
+        synchronized (running) {
+            if (running.get()) {
+                log.warn("Reporter is already running");
+            } else {
+                createStatsd();
```
Comment (alexism) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: `statsd = createStatsd()` for consistency sake?

File: src/main/java/com/airbnb/metrics/ExcludeMetricPredicate.java [hunk 9]
```diff
@@ -29,12 +29,12 @@
 public class ExcludeMetricPredicate implements MetricPredicate {
     private final Logger logger = Logger.getLogger(getClass());
 
-    final String exclude_regex;
+    final String regexFilter;
```
Comment (alexism) [inline src/main/java/com/airbnb/metrics/ExcludeMetricPredicate.java]: `excludeRegex` ?

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java [hunk 10]
```diff
@@ -1,21 +1,19 @@
-/**
+/*
  *
+ * Copyright (c) 2015. Jun He jun.he@airbnb.com
```
Comment (liukai) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: Not a copyright expert, but the copyright owner should be "Airbnb", right?

**Comment by alexism** [comment]
let's simplify the `MetricDimensionOptions` stuff:
1. by default let's send _all_ dimension. that will remove the need for `MetricDimensionOptions.defaultValue`
2. let's use the same dimension names as [this](https://github.com/coursera/metrics-datadog/blob/2.x-maintenance/src/main/java/com/yammer/metrics/reporting/DatadogReporter.java#L212-L241) "p99", "count", etc
3. the various dimensions could be in a separate public `enum Dimension` class. The `StatsDReporter` constructor would receive an `EnumSet` instead of a `MetricDimensionOptions`

File: build.gradle (hunk 1 above)
Comment (jun-he) [inline build.gradle:35]: Yeah, change it to StatsD Support for Kafka Metrics (kafka version: 0.8.1 and 0.8.2)

File: build.gradle (hunk 2 above)
Comment (jun-he) [inline build.gradle:43]: Yes, we need it and it has been included in build.gradle
compile 'com.indeed:java-dogstatsd-client:2.0.10'

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java (hunk 3 above)
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java:48]: It is filter to exclude metrics.  Rename it back.

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java (hunk 4 above)
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: Fix it.  Thanks.

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java (hunk 5 above)
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: Fix them.  Thanks.

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java (hunk 6 above)
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: Rename it.  Thanks.

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java (hunk 8 above)
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: Changed.  Thanks.

File: src/main/java/com/airbnb/metrics/ExcludeMetricPredicate.java (hunk 9 above)
Comment (jun-he) [inline src/main/java/com/airbnb/metrics/ExcludeMetricPredicate.java]: Changed.  Thanks.

File: src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java (hunk 10 above)
Comment (jun-he) [inline src/main/java/com/airbnb/kafka/KafkaStatsdMetricsReporter.java]: Change all copyright owner to Airbnb.com

**Comment by jun-he** [comment]
I think it is OK to keep some parts of current design of MetricDimensionOptions.
1) EnumSet is a good idea, change the code accordingly.
2) default to ALL may send too many info for a single metric.  The current design is OK.
3) "p99" is not easy to understand for the end customer.  The current design is OK.

**Comment by alexism** [comment]
Thanks Jun, it's looking good!
- let's simplify and get rid of `MetricDimensionOptions`
- let's standardize on `p99`, etc. This is what is used in other reporters. It will make our metrics consistent.
- `Parser.getTags` is a bit confusing and not thread safe. `com.airbnb.metrics.StatsDReporter#run` could be invoked concurrently. So,
  - let's remove `Parser.getTags`
  - maybe rename `Parser.parse` into `Parser.parseTags` and return the tags
  - remove `Parse.parse` from `#run()`, and invoke it from the `process*` methods
    For example:

```
    public void processCounter(MetricName metricName, Counter counter, Long context) throws Exception {
        statsd.gauge(parser.getName(), counter.count(), parser.parseTags());
    }
```

**Comment by alexism** [comment]
I noticed that the files are indented with 4 spaces. Standard is 2.

File: src/main/java/com/airbnb/metrics/Parser.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        return sanitizeName(metricName, metricName.getName(), "");
+    }
+
+    String sanitizeName(MetricName metricName, String suffix) {
+        return sanitizeName(metricName, metricName.getName(), suffix);
+    }
+
+    //keep it similar as those in com.airbnb.kafka.KafkaStatsdMetricsReporter
+    String sanitizeName(MetricName metricName, String name, String suffix) {
+        return new StringBuilder(128)
```
Comment (alexism) [inline src/main/java/com/airbnb/metrics/Parser.java]: this is a duplicate of `com.airbnb.metrics.MetricNameFormatter#format`.
let's keep it DRY

File: src/main/java/com/airbnb/metrics/ParserForTagInMBeanName.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+        Pattern p = tagRegexMap.get(metricName.getType());
+        if (p != null && !p.matcher(metricName.getMBeanName()).matches()) {
+            name = sanitizeName(metricName, SUFFIX_FOR_ALL);
+        } else {
+            name = sanitizeName(metricName);
+        }
+        tags = getTags(metricName);
+    }
+
+    String[] getTags(MetricName metricName) {
```
Comment (alexism) [inline src/main/java/com/airbnb/metrics/ParserForTagInMBeanName.java]: this method has the name has the `Parser#getTags` (in the super class). It's confusing.
Also let's make it `private` or `protected`

Comment (alexism) [inline src/main/java/com/airbnb/metrics/ParserForTagInMBeanName.java]: This is the critical part. We should have thorough unit tests on these lines.

File: src/main/java/com/airbnb/metrics/Parser.java (hunk 11 above)
Comment (jun-he) [inline src/main/java/com/airbnb/metrics/Parser.java]: I moved sanitizeName to formatter.  They are quite similar.  The main difference is that format will include scope but sanitizename does not include scope in the name.

File: src/main/java/com/airbnb/metrics/ParserForTagInMBeanName.java (hunk 12 above)
Comment (jun-he) [inline src/main/java/com/airbnb/metrics/ParserForTagInMBeanName.java]: Thanks for the comments.  Add more tests and fixed method name.

**Comment by jun-he** [comment]
/cc @alexism @liukai 
Can you review it and let me know your comments?  Thanks!
Hopefully, we can merge PR this week. :-)

**Pull request merged**